src/shared_realm.cpp

```diff
@@ -554,5 +555,110 @@ util::Optional<int> Realm::file_format_upgraded_from_version() const
+{
+    if (is_awaiting_import()) {
+        m_coordinator->get_realm()->m_shared_group->unpin_version(VERSION_CAST(m_version_id));
+        m_coordinator = nullptr;
+    }
+}
+
+Realm::HandoverPackage Realm::package_for_handover(std::vector<AnyThreadConfined> objects_to_hand_over)
+{
+    verify_thread();
+    if (is_in_transaction()) {
+        throw InvalidTransactionException("Cannot package handover during a write transaction.");
+    }
+
+    HandoverPackage handover;
+    auto version_id = m_shared_group->pin_version();
+    handover.m_version_id = VERSION_CAST(version_id);
+    handover.m_coordinator = m_coordinator;
+    // Since `m_coordinator` is used to determine if we need to unpin when destroyed,
+    // `m_coordinator` should only be set after `pin_version` succeeds in case it throws.
+
+    handover.m_objects.reserve(objects_to_hand_over.size());
+    for (auto &object : objects_to_hand_over) {
+        REALM_ASSERT(object.get_realm().get() == this);
+        handover.m_objects.push_back(object.export_for_handover());
+    }
+
+    return handover;
+}
+
+std::vector<AnyThreadConfined> Realm::accept_handover(Realm::HandoverPackage handover)
+{
+    REALM_ASSERT(handover.is_awaiting_import()); // Enforced by move semantics
+    auto unpin_version = util::make_scope_exit([&]() noexcept {
+        m_shared_group->unpin_version(VERSION_CAST(handover.m_version_id));
+        handover.m_coordinator = nullptr;
+    });
+
+    verify_thread();
+    if (is_in_transaction()) {
+        throw InvalidTransactionException("Cannot accept handover during a write transaction.");
+    }
+
+    // Ensure we're on the same verison as the handover
+    if (!m_group) {
+        // A read transaction doesn't yet exist, so create at the handover version
+        m_group = &const_cast<Group&>(m_shared_group->begin_read(VERSION_CAST(handover.m_version_id)));
+    }
+    else {
+        auto current_version = m_shared_group->get_version_of_current_transaction();
+
+        if ((SharedGroup::VersionID)VERSION_CAST(handover.m_version_id) <= current_version) {
+            // The handover is behind, so advance it to our version
+            handover.advance_to_version(VERSION_CAST(current_version));
+        } else {
+            // We're behind, so advance to the handover's version
+            transaction::advance(*m_shared_group, m_binding_context.get(),
+                                 m_config.schema_mode, VERSION_CAST(handover.m_version_id));
+            m_coordinator->process_available_async(*this);
+        }
+    }
+
+    std::vector<AnyThreadConfined> objects;
+    objects.reserve(handover.m_objects.size());
+    for (auto &object : handover.m_objects) {
+        objects.push_back(std::move(object).import_from_handover(shared_from_this()));
+    }
+
+    // Avoid weird partial-refresh semantics when importing old packages by always refreshing
+    transaction::advance(*m_shared_group, m_binding_context.get(), m_config.schema_mode);
```

Comment:
This is slightly more efficient since it doesn't have to perform a bunch of checks again, but it might not be worth the added code complexity. Is that your reasoning?